\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}

\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses
\usepackage{ifthen}  % Conditional
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}

\def\includegraphic{}
\def\includegraphics{}


\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm}
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%Review style settings
\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

\begin{document}
\begin{bmcformat}


\title{Next generation sequencing information management and analysis
  system for Galaxy}

\author{Brad A Chapman\correspondingauthor$^{1}$%
       \email{chapman@molbio.mgh.harvard.edu}%
      \and
         Mark Borowsky$^1$%
         \email{borowsky@molbio.mgh.harvard.edu}%
      }

\address{%
    \iid(1)Department of Molecular Biology, Simches Research Center,%
    Massachusetts General Hospital, Boston, MA 02114, USA
}%

\maketitle

\begin{abstract}
  \paragraph*{Background:} Next generation sequencing technologies
  like Illumina, SOLiD and 454 have provided core facilities with the
  ability to produce large amounts of sequence data. Along with this
  increased output comes the challenge of managing requests and
  samples, tracking sequencing runs, and automating downstream
  analyses.
  \paragraph*{Results:} We approached these challenges by developing a
  sample submission and tracking interface on top of the web-based
  Galaxy data integration platform. On the back end, an automated
  analysis pipeline processes data as it arrives off the sequencer,
  uploading the results back into Galaxy.
  \paragraph*{Conclusions:} The resulting system demonstrates the
  utility of well run community open source analysis platforms.
  It allows small research cores to effectively process high throughput
  data, thus enabling research scientists to quickly assess experimental
  results.
\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

\section*{Background}

The proliferation of next-generation sequencing machines opens up
large scale sequencing to a multitude of new users. Among these are
Bioinformatics core facilities and individual labs, who require
flexible interfaces for entering samples, tracking sequencing,
processing output files and analyzing the resulting data. These
challenges require computational developers to provide data for
biological users in a format that can be readily queried and
visualized.

The Galaxy data integration platform \cite{goecks_galaxy:_2010}
helps ameliorate the challenge of data management for biologists
by providing web-based analysis tools aimed at biological
researchers. By replicating and expanding the tools a computational
user would have available at a UNIX commandline, biologists can
visualize and process the large amount of data produced by next
generation sequencing platforms. For instance, large multiple gigabyte
alignment files from multiple experiments can be viewed directly
in the UCSC genome browser \cite{fujita_ucsc_2011} from Galaxy,
allowing biologists to immediately view their sequencing data
and assess experimental success. In combination with the analysis
tools and visualization options, Galaxy enables reproducible research
and collaboration between computational scientists and non-programmers by
incorporating multi-tool workflows and sharing of computational
results.

A number of bioinformatics tools have emerged to help with handling
high throughput sequencing data. Sequence alignment programs such as
Bowtie \cite{langmead_ultrafast_2009} and BWA \cite{li_fast_2009}
handle finding read locations within sequenced genomes. Toolkits such
as samtools \cite{li_sequence_2009} and Picard \cite{_picard_????}
allow manipulation and querying of resulting alignment files. Larger
tool suites such as the Genome Analysis Toolkit
\cite{mckenna_genome_2010} work with aligned data to characterize
genetic variation.

A common theme of all these toolkits is a requirement for developing
automated workflows to drive them. For instance, a variant calling
pipeline starting from sequencing data requires more than a dozen
individual programs. These automation requirements are becoming
even more clear with the increase in multiplexed sequencing samples,
which require deconvolution of barcodes prior to starting an analysis
workflow.

The solution to these challenges is capturing experimental data,
including barcodes and run data, in a structured format. Moving from
Excel spreadsheet and lab notebook tracking to this structured
platform allows development of the necessary automation, providing
data to researchers in a timely organized manner. By embedding the
results within the existing Galaxy toolkit, results are immediately
available for analysis by biologists.

Here we describe our fully integration solution that tackles these
issues and takes full advantage of existing software and
infrastructure. The resulting tools effectively improve the
challenging interactions between computational and laboratory
researchers, enabling scientific work.

\section*{Implementation}

Existing sample tracking

Researcher targeted user interface

Automated Illumina integration and back end processing

Distribution of files for visualization and additional analysis

\section*{Results and Discussion}

\subsection*{Enabling research}

\subsection*{Collaborative development}

\subsection*{Bioinformatics core efficiency}

\section*{Conclusions}

\section*{Availability and requirements}

\section*{Authors contributions}

\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}

{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{nglims_galaxy} }  % Bibliography file (usually '*.bib' )

\ifthenelse{\boolean{publ}}{\end{multicols}}{}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
  \subsection*{Figure 1 - Sample figure title}
      A short description of the figure content
      should go here.

  \subsection*{Figure 2 - Sample figure title}
      Figure legend text.



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}
  \subsection*{Table 1 - Sample table title}
    Here is an example of a \emph{small} table in \LaTeX\ using  
    \verb|\tabular{...}|. This is where the description of the table 
    should go. \par \mbox{}
    \par
    \mbox{
      \begin{tabular}{|c|c|c|}
        \hline \multicolumn{3}{|c|}{My Table}\\ \hline
        A1 & B2  & C3 \\ \hline
        A2 & ... & .. \\ \hline
        A3 & ..  & .  \\ \hline
      \end{tabular}
      }
  \subsection*{Table 2 - Sample table title}
    Large tables are attached as separate files but should
    still be described here.



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.

\end{bmcformat}
\end{document}
