\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}

\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses
\usepackage{ifthen}  % Conditional
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}

\def\includegraphic{}
\def\includegraphics{}


\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm}
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%Review style settings
\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

\begin{document}
\begin{bmcformat}


\title{Next generation sequencing information management and analysis
  system for Galaxy}

\author{Brad A Chapman\correspondingauthor$^{1}$%
       \email{chapman@molbio.mgh.harvard.edu}%
      \and
         Mark Borowsky$^1$%
         \email{borowsky@molbio.mgh.harvard.edu}%
      }

\address{%
    \iid(1)Department of Molecular Biology, Simches Research Center,%
    Massachusetts General Hospital, Boston, MA 02114, USA
}%

\maketitle

\begin{abstract}
  \paragraph*{Background:} Next generation sequencing technologies
  like Illumina, SOLiD and 454 have provided core facilities with the
  ability to produce large amounts of sequence data. Along with this
  increased output comes the challenge of managing requests and
  samples, tracking sequencing runs, and automating downstream
  analyses.
  \paragraph*{Results:} We approached these challenges by developing a
  sample submission and tracking interface on top of the web-based
  Galaxy data integration platform. On the back end, an automated
  analysis pipeline processes data as it arrives off the sequencer,
  uploading the results back into Galaxy.
  \paragraph*{Conclusions:} The resulting system demonstrates the
  utility of well run community open source analysis platforms.
  It allows small research cores to effectively process high throughput
  data, thus enabling research scientists to quickly assess experimental
  results.
\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

\section*{Background}

The proliferation of next-generation sequencing machines opens up
large scale sequencing to a multitude of new users. Among these are
Bioinformatics core facilities and individual labs, who require
flexible interfaces for entering samples, tracking sequencing,
processing output files and analyzing the resulting data. These
challenges require computational developers to provide data for
biological users in a format that can be readily queried and
visualized.

The Galaxy data integration platform \cite{goecks_galaxy:_2010}
helps ameliorate the challenge of data management for biologists
by providing web-based analysis tools aimed at biological
researchers. By replicating and expanding the tools a computational
user would have available at a UNIX commandline, biologists can
visualize and process the large amount of data produced by next
generation sequencing platforms. For instance, large multiple gigabyte
alignment files from multiple experiments can be viewed directly
in the UCSC genome browser \cite{fujita_ucsc_2011} from Galaxy,
allowing biologists to immediately view their sequencing data
and assess experimental success. In combination with the analysis
tools and visualization options, Galaxy enables reproducible research
and collaboration between computational scientists and non-programmers by
incorporating multi-tool workflows and sharing of computational
results.

A number of bioinformatics tools have emerged to help with handling
high throughput sequencing data. Sequence alignment programs such as
Bowtie \cite{langmead_ultrafast_2009} and BWA \cite{li_fast_2009}
handle finding read locations within sequenced genomes. Toolkits such
as samtools \cite{li_sequence_2009} and Picard \cite{_picard_????}
allow manipulation and querying of resulting alignment files. Larger
tool suites such as the Genome Analysis Toolkit
\cite{mckenna_genome_2010} work with aligned data to characterize
genetic variation.

A common theme of all these toolkits is a requirement for developing
automated workflows to drive them. For instance, a variant calling
pipeline starting from sequencing data requires more than a dozen
individual programs. These automation requirements are becoming
even more clear with the increase in multiplexed sequencing samples,
which require deconvolution of barcodes prior to starting an analysis
workflow.

The solution to these challenges is capturing experimental data,
including barcodes and run data, in a structured format. Moving from
Excel spreadsheet and lab notebook tracking to this structured
platform allows development of the necessary automation, providing
data to researchers in a timely organized manner. By embedding the
results within the existing Galaxy toolkit, results are immediately
available for analysis by biologists.

Here we describe our fully integration solution that tackles these
issues and takes full advantage of existing software and
infrastructure. The resulting tools effectively improve the
challenging interactions between computational and laboratory
researchers, enabling scientific work.

\section*{Implementation}

Building an automated system for processing next-generation sequencing
data required two components. The first is a front-end web interface,
tightly integrated with Galaxy, that provides a graphical method to
quickly collect and manage sample data. The second is a back-end set
of scripts which drives analysis based on sample data and reintegrates
the results with the Galaxy front-end for researcher analysis.

\subsection*{Front-end Galaxy interface}

\subsubsection*{Researcher sample entry}

Biologists use a local Galaxy server as an entry point to submit
samples for sequencing. This provides a familiar interface and central
location for both entering sample information and retrieving and
analyzing the sequencing data.

Practically, a user begins by browsing to the sample submission
page. There they are presented with a wizard interface which guides
them through entry of sample details. Multiplexed samples are
supported through a drag and drop interface (Figure 1).

When all samples are entered, the user submits them as a sequencing
project. This includes billing information and a project name to
facilitate communication between the researcher and the core group
about submissions (Figure 2). Users are able to view their submissions
grouped as projects and track the state of constructs. The interface
allows support for additional services associated with sequencing,
like library construction, quantitation and validation. This is a
valuable way for users to track and organize the status of their
samples.

\subsubsection*{Sequencing tracking and management}

Administrators and sequencing technicians have access to additional
functionality to help manage the internal sample preparation and
sequencing workflow. The main sample tracking interface centers around
a set of queues; each queue represents a state that a sample can be
in. Samples move through the queues as they are processed, with
additional information being added to the sample at each step. For
instance, a sample in the ‘Pre-sequencing quantitation’ queue moves to
the ‘Sequencing’ queue once it has been fully quantitated, with
quantitation information entered by the sequencing technician during
the transition.

Assigning samples to flow cells occurs using a drag and drop jQueryUI
interface. The design is flexible to allow for placing samples across
multiple lanes or multiplexing multiple barcoded samples into a single
lane (Figure 3).

\subsubsection*{Viewing sequencing results}

Running a sequencing machine requires careful monitoring of results
and our interface provides several ways to view this data. Raw cluster
and read counts are linked to a list of runs. For higher level
analyses, interactive plots are available for viewing reads over time
and pass rates compared to read density. These allow adjustment of
experimental procedures to maximize useful reads based on current
machine chemistry (Figure 4).

The results of processing sequencing runs -- fastq reads, alignments
files, summary PDFs and other associated files -- are uploaded back
into into Galaxy Data Libraries organized by sample names (Figure
5). Users can download results for offline work, or import them
directly into their Galaxy history for further analysis or display.

\subsection*{Back-end analysis pipeline}

Analysis files are produced by an automated pipeline of scripts
written in the Python programming language. A script run hourly by a
cron job detects runs as they come off the sequencer. Finished run
results are processed into standard fastq files using Illumina
utilities as the start point for specialized processing.

Details about the finished run are passed on to storage and analysis
servers using AMQP messaging. The storage server transfers files from
the sequencing machine for long term archival and backup purposes. The
analysis server transfers fastq files in preparation for detailed
analyses (Figure 6).

Different analysis options are available depending on the sequencing
inputs and experimental goals. The most complete pipeline is for
variant calling, which demonstrates the major features of the system:

\begin{itemize}
  \item De-multiplex lane results into separate samples, combining
    samples run on multiple lanes \cite{cock_biopython:_2009}.
  \item Alignment with Bowtie or BWA
    \cite{langmead_ultrafast_2009,li_fast_2009}.
  \item Generation of alignment and read statistics with Picard, the
    FASTX-toolkit and SolexaQA
    \cite{_picard_????,_fastx-toolkit_????,cox_solexaqa:_2010}.
  \item Preparation of a summary PDF with detailed statistics
    about the run and alignment.
  \item Recalibration and realignment using the Genome Analysis
    Toolkit
    \cite{mckenna_genome_2010,_pysam_????,gautier_intuitive_2010}.
  \item Variant identification with Unified Genotyper from the Genome
    Analysis Toolkit.
  \item Variant effect prediction using snpEff \cite{_snpeff_????}.
\end{itemize}

The pipeline is completely general and driven by open-source tools.
Additional analysis pipelines for experimental approaches like
ChIP-seq or RNA-seq could be readily integrated.

\section*{Results and Discussion}

\subsection*{Collaborative development}

A unique aspect of this project is that it was undertaken
independently by the Bioinformatics core at Massachusetts General
Hospital, with the front-end code being subsequently integrated
into the main Galaxy codebase. This was made possible because of the
strong software development ethos at Galaxy. It is a well organized,
open-source project whose development directions and goals are fully
transparent to the greater programming community.

We hope this type of community code development will become
increasingly common. With the continuing onslaught of next-generation
sequencing data, computational biologists will managing and analyzing
large volumes of experimental data coming out of laboratories. One
productive way to meet this challenge is to better train ourselves to
build off existing frameworks, utilize open-source tools, and make our
code more interoperable.

This work would certainly not have been possible for a developer in a
small core lab without the backbone of functionality provided by
Galaxy.

\subsection*{Enabling biological research}

The largest beneficiary of this work has been the biological
researchers within our department. By establishing a structured intake
system and automated analysis pipeline, retrieving initial results
does not require manual work by bioinformaticians. As a result
the time to expect an analysis of a next-generation sequencing run has
fallen to only the computational time required for the alignments and
post-processing.

We find it an extremely positive development to get e-mails from
biologists hours after a run is done wondering when they can have
their data in Galaxy. The move from a model where processing would
require manual intervention opens up entirely new classes of
experiments and accelerates the time between designing an experiment
and determining whether the data fits the expected hypotheses.

\subsection*{Bioinformatics core efficiency}

A secondary benefit of the automated approach is freeing
bioinformatics core groups from the day-to-day management of
sequencing data. It is easy for a small group to have to move into
firefighting mode as data production overwhelms the ability to
analyze it. With researchers handling data entry and automated systems
taking care of initial processing tasks, computational biologists are
freed to focus on algorithm development, higher level analyses and
specific visualizations. This transformation allows computational team
members to explore more intricate analysis approaches, and overall
makes a core team able to more effectively contribute to research
output.

\section*{Conclusions}

We describe a web based laboratory information system for
next-generation sequencing data. By developing the entry and
management forms inside of Galaxy, researchers get advanced analysis
tools along with sequencing data and analysis results. The structured
data collection of sample and sequencing information allowed the
development of an automated back end system which moves results from
the sequencers through specialized analysis pipelines. The resulting
infrastructure has fundamentally improved our ability to collaborate
with laboratory researchers.

This functionality was uniquely developed by an informal collaboration
between a remote bioinformatics core group and the Galaxy development
team. By combining several open-source toolkits and algorithms, the
resulting system can be readily adopted by other facilities. The
flexibility and maintainability of the code base is designed to make
extending and using this software a joy.

\section*{Availability and requirements}

The front-end entry system is available as part of Galaxy; see
\url{http://getgalaxy.org} for installation details and
\url{https://bitbucket.org/galaxy/galaxy-central/wiki/LIMS/nglims} for
instructions on configuration and activation.

The back-end analysis pipeline is available from our GitHub repository
(\url{https://github.com/chapmanb/bcbb/tree/master/nextgen}) along
with installation instructions.

Both are made freely available for both academic and commercial use
under the permissive MIT license.

\section*{Authors contributions}

BAC conceived and wrote the software. XXX and XXX implemented crucial
Galaxy components and helped with integration. MB oversaw the software
development and XXX oversaw Galaxy development.

\section*{Acknowledgements}

  \ifthenelse{\boolean{publ}}{\small}{}

{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{nglims_galaxy} }  % Bibliography file (usually '*.bib' )

\ifthenelse{\boolean{publ}}{\end{multicols}}{}

\section*{Figures}
\subsection*{Figure 1 - Researcher sample entry}
Entry of sequencing samples is fully integrated with Galaxy. A
wizard interface with automated completion from previous samples
allows quick collection of important details. A drag and drop
interaction makes specifying multiplexed barcode details painless.

\url{http://www.youtube.com/watch?v=HGhNMeEAFV0}

\subsection*{Figure 2 - Project submission and tracking}
Related samples are submitted as a project. The interface allows
researchers to follow the state of the samples through the
preparation and sequencing process for full transparency.

\url{http://www.youtube.com/watch?v=DtQG9IzpoCU}

\subsection*{Figure 3 - Sample management by sequencing technicians}
Technicians doing the sequencing work move samples through multiple
queue states, tracking important sample details. Sequencing samples
are placed on flowcells using a drag-drop interface.

\url{http://www.youtube.com/watch?v=Sjt6y1lbzVI}

\subsection*{Figure 4 - Sequencing results}
Important statistics about sequencing runs are available to
researchers and sequencing technicians. Interactive plots enable
assessment of results over time.

\url{http://www.youtube.com/watch?v=4xrtPXE7Oe8}

\subsection*{Figure 5 - Sequencing data and analysis results}
Results are readily available to users in Galaxy data libraries. This
allows easy import into Galaxy history for analysis and visualization.

TODO: Make a video for this to show importing results into
history.
\url{http://bcbio.files.wordpress.com/2011/01/data_library_results.png}

\subsection*{Figure 6 - Back end analysis architecture}
Location of detection, storage and analysis servers running in the
back-end pipeline.

TODO: Figure of overall system + Galaxy
\end{bmcformat}
\end{document}
